### PR TITLE
feat: add nano contract info to tx history [6]

### PR DESCRIPTION
**Pull request by alexruzenhack**
This change enables to show nano contract context information on regular transaction list and details. Also, it enables to set marks based on transaction type and make visual differentiation among them.

Look at the approved design https://github.com/HathorNetwork/hathor-wallet-mobile/issues/445. Go to "Flow 4".

### Acceptance Criteria
- should extract TxHistory construction to the model
- should add nano contract information on tx details view
- should change icon to access token info

##### ActionDot to access token info

---

>[!NOTE]
>It depends on:
>- https://github.com/HathorNetwork/hathor-wallet-mobile/pull/453
>- https://github.com/HathorNetwork/hathor-wallet-mobile/pull/454
>- https://github.com/HathorNetwork/hathor-wallet-mobile/pull/432
>- https://github.com/HathorNetwork/hathor-wallet-lib/pull/630

### Security Checklist
- [x] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/components/TxDetailsModal.js [hunk 1]
```diff
@@ -29,11 +30,21 @@ class TxDetailsModal extends Component {
...
+     * }} TxDetailsModal properties
+     */
     const { token, tx, isNFT } = this.props;
     const fullTokenStr = getTokenLabel(token);
     const description = tx.getDescription(token);
     const timestampStr = tx.getTimestampFormat();
     const idStr = getShortHash(tx.txId, 12);
+    const ncId = tx.ncId && getShortHash(tx.ncId, 7);
+    const ncMethod = tx.ncMethod;
+    const ncCallerAddress = tx.ncCaller && getShortContent(tx.ncCaller.base58, 7);
```
Comment (alexruzenhack) [inline src/components/TxDetailsModal.js:47]: @pedroferreira1 That is the motivation to get nano contract basic information in `IHistoryTx` as I pointed at this [comment](https://github.com/HathorNetwork/hathor-wallet-lib/pull/613#discussion_r1513496110).

**Review by andreabadesso — APPROVED** [review]

**Comment by anafigueired** [comment]
The TokensHeader component is not using the new one.

Besides that, everything meets the Design accordingly.

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/components/ActionDot.icon.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/ActionDot.icon.js:1]: No, it is because the parent PR doesn't have the icons, but the saga effect to load the nano contract history. The only way to solve this puzzle is by merging the parent PR.

File: src/components/TransactionStatusLabel.component.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
+/**
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/TransactionStatusLabel.component.js]: No. No rebase missing, they are ortogonal.

File: src/components/HathorFlatList.component.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/**
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/HathorFlatList.component.js]: No. No rebase missing, they are ortogonal.

File: src/components/SimpleButton.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ import {
 import { PRIMARY_COLOR } from '../constants';
 
 const SimpleButton = (props) => {
+  const { children } = props;
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/SimpleButton.js]: Done

File: src/components/TxDetailsModal.js [hunk 6]
```diff
@@ -50,18 +94,28 @@ class TxDetailsModal extends Component {
...
+                <View>
+                  <ListItem title={t`Token`} text={fullTokenStr} />
+                  <ListItem title={t`Description`} text={description} />
+                  <ListItem title={t`Date & Time`} text={timestampStr} />
+                  <ListItem title={t`Transaction ID`} text={txIdComponent} />
+                  {isNc && <ListItem title={t`Blueprint Method`} text={ncMethod} />}
+                  {isNc && <ListItem title={t`Nano Contract ID`} text={ncIdComponent} />}
+                  {isNc && <ListItem title={t`Nano Contract Caller`} text={ncCallerAddrComponent} />}
+                  {isNc && <ListItem title={t`Status`} text={txStatus} />}
+                  {isNc && <PublicExplorerListButton txId={shortNcId} title={t`Nano Contract`} />}
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/TxDetailsModal.js]: Translation updated

File: src/components/ActionDot.icon.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/components/ActionDot.icon.js]: I don't understand, the component PR you linked here also doesn't have the icons, how would it solve this?

[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/components/ActionDot.icon.js]: Nevermind, re-checked the code and the icons were moved

File: locale/da/texts.po [hunk 7]
```diff
@@ -12,56 +12,101 @@ msgstr ""
...
+msgstr ""
+
+#. *
+#.  * Map version type to info that should assist on token list visualization.
+#.  *
+#.  * @readonly
+#.  
+#: src/models.js:36
+msgid "Nano Contract Transaction"
+msgstr "Ingen transaktioner"
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply locale/da/texts.po:58]: I think it was translated automatically...

File: src/components/Icons/CircleCheck.icon.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
+/**
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/Icons/CircleCheck.icon.js:1]: It needs a rebase

File: src/components/PublicExplorerListButton.js [hunk 9]
```diff
@@ -7,15 +7,21 @@ import { ListButton } from './HathorList';
 import { COLORS } from '../styles/themes';
 import { combineURLs } from '../utils';
 
-export function PublicExplorerListButton(props) {
-  const { txId } = props;
+const DEFAULT_TITLE = t`Public Explorer`;
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/PublicExplorerListButton.js:10]: Sure!

File: src/components/TransactionStatusLabel.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * @description
+ * This component was devised to be used in Nano Contract context
+ * to provide a visual indication for transaction "status".
+ * @param {Object} param
+ * @param {boolean} param.hasFirstBlock It indicates if a transaction has a first block
+ * @param {boolean} param.isVoided Transaction's void flag
+ */
+export const TransactionStatusLabel = ({ hasFirstBlock, isVoided = false }) => {
```
[the remark this answers is not in the document]
Comment (alexruzenhack) [reply src/components/TransactionStatusLabel.js:94]: Oh! I removed it in the previous PRs because hathor-core is not prepared yet for this.

File: locale/pt-br/texts.po [hunk 11]
```diff
@@ -1180,10 +1221,30 @@ msgstr "Processando"
...
 msgstr "Inválida"
 
-#: src/components/TxDetailsModal.js:60
+#: src/components/TxDetailsModal.js:101
 msgid "Description"
 msgstr "Descrição"
 
+#: src/components/TxDetailsModal.js:103
+msgid "Transaction ID"
+msgstr "Transação"
```
Comment (andreabadesso) [inline locale/pt-br/texts.po:1230]:
```suggestion
msgstr "ID da transação"
```

File: locale/pt-br/texts.po [hunk 12]
```diff
@@ -1180,10 +1221,30 @@ msgstr "Processando"
...
+
+#: src/components/TxDetailsModal.js:104
+msgid "Blueprint Method"
+msgstr "Método Blueprint"
+
+#: src/components/TxDetailsModal.js:105
+msgid "Nano Contract ID"
+msgstr ""
```
Comment (andreabadesso) [inline locale/pt-br/texts.po:1238]:
```suggestion
msgstr "ID do Nano Contract"
```

File: locale/pt-br/texts.po [hunk 13]
```diff
@@ -1180,10 +1221,30 @@ msgstr "Processando"
...
+
+#: src/components/TxDetailsModal.js:106
+msgid "Nano Contract Caller"
+msgstr ""
```
Comment (andreabadesso) [inline locale/pt-br/texts.po:1242]: This one is interesting, what do you think about "Caller do Nano Contract"?

File: src/reducers/reducer.js [hunk 14]
```diff
@@ -11,6 +11,7 @@ import { INITIAL_TOKENS, DEFAULT_TOKEN, PUSH_API_STATUS, FEATURE_TOGGLE_DEFAULTS
 import { types } from '../actions';
 import { TOKEN_DOWNLOAD_STATUS } from '../sagas/tokens';
 import { WALLET_STATUS } from '../sagas/wallet';
+import { TxHistory } from '../models';
```
Comment (andreabadesso) [inline src/reducers/reducer.js:14]: Unused

File: src/sagas/tokens.js [hunk 15]
```diff
@@ -249,7 +253,7 @@ export function* fetchTokenMetadata({ tokenId }) {
       tokenId,
     });
     // eslint-disable-next-line
```
Comment (andreabadesso) [inline src/sagas/tokens.js:255]: Why do we need to disable eslint here? I think this can be removed

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: locale/pt-br/texts.po [hunk 16]
```diff
@@ -1180,10 +1221,30 @@ msgstr "Processando"
...
+msgid "Blueprint Method"
+msgstr "Método Blueprint"
+
+#: src/components/TxDetailsModal.js:105
+msgid "Nano Contract ID"
+msgstr "ID do Nano Contract"
+
+#: src/components/TxDetailsModal.js:106
+msgid "Nano Contract Caller"
+msgstr "Acionador do Nano Contract"
```
Comment (andreabadesso) [inline locale/pt-br/texts.po:1242]: What would be the best person to discuss this translation with? Campana?

I don't like this translation, but I don't have anything better to suggest, so I'm going to approve this PR

**Review by andreabadesso — APPROVED** [review]

File: locale/pt-br/texts.po (hunk 16 above)
Comment (pedroferreira1) [reply locale/pt-br/texts.po:1242]: I don't like it either. While we wait for the best names from the other areas, I think we can call "Endereço que assinou o Nano Contract", even if it's big, I think it's better.

What do you think?

File: src/models.js [hunk 17]
```diff
@@ -8,13 +8,100 @@
...
+ * @readonly
+ */
+const VersionInfo = {
+  [VersionType.blockVersion]: { symbol: 'b', label: t`Block` },
+  [VersionType.transactionVersion]: { symbol: 'tx', label: t`Transaction` },
+  [VersionType.createTokenTransactionVersion]: { symbol: 'tk', label: t`Create Token Transaction` },
+  [VersionType.mergedMinedBlockVersion]: { symbol: 'mb', label: t`Merged Mined Block` },
+  [VersionType.nanoContractTransaction]: { symbol: 'nc', label: t`Nano Contract Transaction` },
+};
+
```
Comment (pedroferreira1) [inline src/models.js:37]: I think we should use the method we already have in the lib: https://github.com/HathorNetwork/hathor-wallet-lib/blob/master/src/utils/transaction.ts#L629

The explorer and desktop wallet use it, and it's better to have a single source of information, so it's easier to maintain.

File: src/models.js [hunk 18]
```diff
@@ -8,13 +8,100 @@
...
+     */
     this.isVoided = isVoided;
+    /**
+     * @type {number}
+     */
+    this.version = version;
+    /**
+     * @type {'processing'|'finished'}
+     */
+    this.processingStatus = processingStatus
```
Comment (pedroferreira1) [inline src/models.js:92]: Do we need this? This is an internal status of the lib.

File: locale/pt-br/texts.po [hunk 19]
```diff
@@ -1180,10 +1221,30 @@ msgstr "Processando"
...
 msgid "Description"
 msgstr "Descrição"
 
+#: src/components/TxDetailsModal.js:103
+msgid "Transaction ID"
+msgstr "ID da Transação"
+
+#: src/components/TxDetailsModal.js:104
+msgid "Blueprint Method"
+msgstr "Método Blueprint"
```
Comment (pedroferreira1) [inline locale/pt-br/texts.po:1234]:
```suggestion
msgstr "Método do Blueprint"
```

File: src/utils.js [hunk 20]
```diff
@@ -31,6 +31,13 @@ export const getShortHash = (hash, length = 4) => (
   `${hash.substring(0, length)}...${hash.substring(64 - length, 64)}`
 );
 
+/**
+ * It short any string content without length bound.
```
Comment (pedroferreira1) [inline src/utils.js:35]: Missing parameters in the docstring

File: src/components/PublicExplorerListButton.js [hunk 21]
```diff
@@ -7,15 +14,21 @@ import { ListButton } from './HathorList';
...
+
+export function PublicExplorerListButton({ txId, title }) {
   const explorerIcon = <Image source={icShareActive} width={24} height={24} />;
   const baseExplorerUrl = useSelector((state) => state.networkSettings.explorerUrl);
   const txUrl = `transaction/${txId}`;
   // XXX: maybe we should have this on the constants or utils to check the network
   const explorerLink = combineURLs(baseExplorerUrl, txUrl);
 
   return (
-    <ListButton title={t`Public Explorer`} button={explorerIcon} onPress={() => { Linking.openURL(explorerLink); }} titleStyle={{ color: COLORS.textColorShadow }} isLast />
```
Comment (pedroferreira1) [inline src/components/PublicExplorerListButton.js:19]: Why did you remove the `isLast` parameter?

File: src/utils.js [hunk 22]
```diff
@@ -329,22 +336,16 @@ export const changePin = async (wallet, oldPin, newPin) => {
...
-    txId: element.txId,
-    timestamp: element.timestamp,
-    balance: element.balance,
-    // in wallet service this comes as 0/1 and in the full node comes with true/false
-    isVoided: Boolean(element.voided),
-    tokenUid: token
-  };
-  return new TxHistory(data);
-};
+export const mapToTxHistory = (tokenUid) => (rawTxHistory) => (
```
Comment (pedroferreira1) [inline src/utils.js:346]: What's the benefit of having this instead of `export const mapToTxHistory = (tokenUid, rawTxHistory)`?

File: src/sagas/tokens.js [hunk 23]
```diff
@@ -133,7 +137,7 @@ function* fetchTokenHistory(action) {
     }
 
     const response = yield call(wallet.getTxHistory.bind(wallet), { token_id: tokenId });
-    const data = response.map((txHistory) => mapTokenHistory(txHistory, tokenId));
+    const data = response.map(mapToTxHistory(tokenId));
```
Comment (pedroferreira1) [inline src/sagas/tokens.js:140]: Why did you change this here but in the actions is different?

image

File: src/components/TokenSelect.js [hunk 24]
```diff
@@ -50,7 +51,6 @@ const TokenSelect = (props) => {
 
     return (
       <TouchableHighlight
-        style={index === 0 ? styles.firstItemWrapper : null}
```
Comment (pedroferreira1) [inline src/components/TokenSelect.js:53]: Why did you remove this style? This changes other places as well

File: src/components/TxDetailsModal.js [hunk 25]
```diff
@@ -50,18 +87,27 @@ class TxDetailsModal extends Component {
         onBackButtonPress={this.props.onRequestClose}
         onBackdropPress={this.props.onRequestClose}
         style={this.style.modal}
+        propagateSwipe
```
Comment (pedroferreira1) [inline src/components/TxDetailsModal.js:90]: What's this?

File: src/components/TxDetailsModal.js [hunk 26]
```diff
@@ -50,18 +87,27 @@ class TxDetailsModal extends Component {
...
             <BalanceView tx={tx} token={token} isNFT={isNFT} />
-            <View>
-              <ListItem title={t`Token`} text={fullTokenStr} />
-              <ListItem title={t`Description`} text={description} />
-              <ListItem title={t`Date & Time`} text={timestampStr} />
-              <ListItem title={t`ID`} text={txIdComponent} />
-              <PublicExplorerListButton txId={tx.txId} />
-            </View>
+            <ScrollView>
+              <TouchableWithoutFeedback>
```
Comment (pedroferreira1) [inline src/components/TxDetailsModal.js:97]: Why do you need this here?

File: src/screens/MainScreen.js [hunk 27]
```diff
@@ -201,10 +201,9 @@ class MainScreen extends React.Component {
     const renderRightElement = () => {
       if (this.props.selectedToken.uid !== hathorConstants.HATHOR_TOKEN_CONFIG.uid) {
         return (
-          <SimpleButton
-            icon={infoIcon}
-            onPress={this.tokenInfo}
-          />
+          <SimpleButton onPress={this.tokenInfo}>
+            <ActionDot />
```
Comment (pedroferreira1) [inline src/screens/MainScreen.js:205]: What's this change? Do you have a screen with this in the description?

File: src/screens/MainScreen.js [hunk 28]
```diff
@@ -454,37 +433,33 @@ class TxListItem extends React.Component {
...
-            {image}
-            <View style={style.middleView}>
-              <Text style={style.description}>{description}</Text>
-              <Text style={style.timestamp}>{timestamp}</Text>
-            </View>
-            <Text style={style.balance} numberOfLines={1}>{balanceStr}</Text>
+      <TouchableHighlight
+        onPress={() => this.onItemPress(item)}
+        underlayColor={COLORS.primaryOpacity30}
+      >
```
Comment (pedroferreira1) [inline src/screens/MainScreen.js:452]: Why did you remove the `touchStyle` and added this `underlayColor`?

File: src/models.js (hunk 17 above)
Comment (alexruzenhack) [reply src/models.js]: I need a map to the translation, the lib doesn't give me a translation.

File: src/utils.js (hunk 22 above)
Comment (alexruzenhack) [reply src/utils.js:346]: Conciseness. I think this is simple to understand and improve readability.

File: src/components/TokenSelect.js (hunk 24 above)
Comment (alexruzenhack) [reply src/components/TokenSelect.js:53]: We don't need to differentiate the first or the last item anymore.

File: locale/pt-br/texts.po (hunk 16 above)
Comment (BrunoCampana) [reply locale/pt-br/texts.po]: @alexruzenhack 

At the moment, in our documentation about nano contracts, this field was named "caller", you can see in the diagram at:
https://beta-test.docs.hathor.network/explanations/features/nano-contracts/how-it-works#transactions

However, please note that this name was provided for two reasons:
- This is how Marcelo named it.
- It relates to who is indeed calling the contract.

That said, although this nomenclature is adequate for the documentation, it might not be suitable in terms of UI/UX. Because I don't know the overall terminology of this and also other features in our UIs, is difficult to suggest something at this moment.

Comment (alexruzenhack) [reply locale/pt-br/texts.po]: Hi, @BrunoCampana, thank you to add more light in this case. I got it!

Regarding the translation, do you do have a suggestion?

I found a better word for the translation so far: "Signatário"; see [definition](https://www.dicio.com.br/signatario/). What do you guys say about this one?

File: src/models.js (hunk 17 above)
Comment (pedroferreira1) [reply src/models.js]: We have many places where we get strings from the lib, e.g., error messages that are not translated. 

We already discussed ways of handling this. I don't think the best way is to duplicate what we have in the lib, this decreases code maintainability a lot.

For now, I prefer keeping these strings not translated and getting them from a single source of information for this kind of thing, which is the lib method.

We need to discuss this in an issue for sure, can you create one for this?

Comment (alexruzenhack) [reply src/models.js]: Sure! I will open an issue for this.

Comment (alexruzenhack) [reply src/models.js]: Here the issue: https://github.com/HathorNetwork/internal-issues/issues/292

Comment (alexruzenhack) [reply src/models.js]: Removed translation mapping: https://github.com/HathorNetwork/hathor-wallet-mobile/pull/434/commits/ccd0f1fad0300196851d1c2287829e8600789d1d

**Review by pedroferreira1 — APPROVED** [review]
